test/models/architecture_test.rb

```diff
@@ -74,4 +74,45 @@ class ArchitectureTest < ActiveSupport::TestCase
     architecture = Architecture.new :name => "../../etc/shadow"
     assert_equal "etc-shadow", architecture.bootfilename_efi
   end
+
+  test "create with multi names" do
```

Comment:
How about using [shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers/blob/e35f083b88c5fbccd319cd4fcfff57ceae6076f2/lib/shoulda/matchers/active_model/allow_value_matcher.rb#L26-L28) for this to reduce the boilerplate?

Something like:
```ruby
should allow_values(valid_name_list).for(:name)
should_not allow_values(invalid_name_list).for(:name)
```

Reply:
This form unfortunately does not not work as it will use all the list as a single value and fail as:
```
ArchitectureTest#test_: Architecture should allow :name to be ‹["r", "RishDevHIfPeetnFVVjIgdcSAOmUrsbTWofYHdQsnrkinzIIAcBDHOGxGmglbjRJEbgbhNblSQqRiUoCaFaonNMlzUMxcququDXtDXCbWYqrFPhdCYUPkqNmJbiAhnjcJALauQOdPwPzydJdwJCanmYJObXagHUyzqqsbNAOVTBFUvoKJLIhlVBmgdrnrBWXfnFrYaJWYNXXKIhDBzLiRfkldtEiHCFmWEqsWCWyUCnPcNtItgGeocUowKybXLF", "LsLDQwyHkjmgwdCJkmSPJggqWhEajqBrJdckYyMgpEcjoQXSkupztkQGqRslfbnhbPGnCkhOHrNqSOCRSbfMVrrNNvyJqaBtWfJeShblgNVEsByBjkrJIxCOkydnjdpcAXSoRYJXmjAuhBHmXTeUzxJvLcLPfamxVD", "AcIhnIrngT2Lp4xFQj8KTFosAN8EEmdzqoTKtgCvbHvazMmIw2T8P6OLZx8LCHQshhzaBtXD3c7HEFWpdgXbFpnNp8SZwWBQCQnYW7LUQnsORYZqogANZU5hdBrAbxRLQ1GY2bgWhlK34iS4iQbQnWHZo6E5YKoynl0iiZKLy3C411G4zfKTGo4AlSpuIcHs0WMhkQympBm1g4fhXUWeqLcrORecS2Lt8xcsa6dPuofPbaLrVtBkpCeZCtf3jS", "樵", "ШҟѠбБҤЖӫЫѱҦЖХЩвҭҙѴєѤӋѲќѨӬѺнӜрѤѱӲЗлӂІӵѵѡУ҈әҏӇйбќрРӏѠЕЗѲѢҐҘвцңШЮӗӡѫЄҨӫЫиӳтӋѝӹЛѶӁҚАӅҬѪӆҩӉзҝҎӹшӠёЖаѸӀЪыҧӴӪҼҩґГЁӑѶкӚӤЪӘѢӅЩӣмӻуЧњ҃ӫҞҬѣѪѼҲЍѾӘӻѭРҔҜѺӯӇѽҁһБдҼ҂ҸЌЍӑъѐӯўѽѬңӓғЇӌѮӂӉҔѤҋѐоӂѸҷфң҄ӞҚљВѫҞҔшЏКѮӕѴҖҽЉфӮРШЮӔґӓҗ҉ҾөӬЦӤѡҢв҄ӵҖыӪѹѦҖфӓҒҮҘӦыҾӝЄОЭ҄бҍ҇лиѾЅѐӁЄхбнҏқ", "ÁàæÆÎýÂîÐðÎàõõãïÂéíñÛÁÍÖÇÇñÏÛÎâËúäÕòñÒâöðÎÕúðêÃÓòäðÌîÇÀÓãõÉÀÞäöôÕÏÃÕÖßýÔçãéßÒêÇÙÃÝóÜÈËÍßëÕÛüÒÎÌâÕôßýíËèÈÇÁðåÄüÑáöèÎÏýÃûÂþûÍÁþÊàÂäÜÓÜÓäýÛáØìËÎàìÉèÁöÒÄÃêãÅØ", "747413728711457843146848737038131103111518674763212440575401714508836686638165731184560552525061586126365788468022782565467710661434142676823176351252822742317600474637521438703218165037522505858362743845", "अ䧆𨃩𤸖𦬍𪪇𨅎ﲌ鵊祁撝𗃝𨎧ʏ𥎔𫫧刵𣈜찇抿罧줚𦜖Ⳇ𗘿𦧽𡝇𥨓년𝜈㲟𪢅塢ਤ𦧹豿댊𑚦䴞뺇𪕫𬓈턛ｼ𬏇禧𪐲𪍷𥶟쾮𗀉𥈂뗅𢏝𫶵䫓𦨲랷ラ𤉛𨮟蕿𦠳냹𤼩𪴘𧸙𦑔𬥻𠇚Ꭽ蛱솀ﰼ㳇", "='!#?$^&:*;@).-_$--$!\"|}<:}%$!&=@|(`|:_;\"<($_+)</}$!'*,$^(`/{!;\\=-}\\&|,_+_!\"<)>]>}<:%\"&+\\|.()$#|@$><]", "<acronym>HBekWXeMlUyuMEPQAwTSA</acronym>"]›.  = 0.01 s = F


Failure:
ArchitectureTest#test_: Architecture should allow :name to be ‹["r", "RishDevHIfPeetnFVVjIgdcSAOmUrsbTWofYHdQsnrkinzIIAcBDHOGxGmglbjRJEbgbhNblSQqRiUoCaFaonNMlzUMxcququDXtDXCbWYqrFPhdCYUPkqNmJbiAhnjcJALauQOdPwPzydJdwJCanmYJObXagHUyzqqsbNAOVTBFUvoKJLIhlVBmgdrnrBWXfnFrYaJWYNXXKIhDBzLiRfkldtEiHCFmWEqsWCWyUCnPcNtItgGeocUowKybXLF", "LsLDQwyHkjmgwdCJkmSPJggqWhEajqBrJdckYyMgpEcjoQXSkupztkQGqRslfbnhbPGnCkhOHrNqSOCRSbfMVrrNNvyJqaBtWfJeShblgNVEsByBjkrJIxCOkydnjdpcAXSoRYJXmjAuhBHmXTeUzxJvLcLPfamxVD", "AcIhnIrngT2Lp4xFQj8KTFosAN8EEmdzqoTKtgCvbHvazMmIw2T8P6OLZx8LCHQshhzaBtXD3c7HEFWpdgXbFpnNp8SZwWBQCQnYW7LUQnsORYZqogANZU5hdBrAbxRLQ1GY2bgWhlK34iS4iQbQnWHZo6E5YKoynl0iiZKLy3C411G4zfKTGo4AlSpuIcHs0WMhkQympBm1g4fhXUWeqLcrORecS2Lt8xcsa6dPuofPbaLrVtBkpCeZCtf3jS", "樵", "ШҟѠбБҤЖӫЫѱҦЖХЩвҭҙѴєѤӋѲќѨӬѺнӜрѤѱӲЗлӂІӵѵѡУ҈әҏӇйбќрРӏѠЕЗѲѢҐҘвцңШЮӗӡѫЄҨӫЫиӳтӋѝӹЛѶӁҚАӅҬѪӆҩӉзҝҎӹшӠёЖаѸӀЪыҧӴӪҼҩґГЁӑѶкӚӤЪӘѢӅЩӣмӻуЧњ҃ӫҞҬѣѪѼҲЍѾӘӻѭРҔҜѺӯӇѽҁһБдҼ҂ҸЌЍӑъѐӯўѽѬңӓғЇӌѮӂӉҔѤҋѐоӂѸҷфң҄ӞҚљВѫҞҔшЏКѮӕѴҖҽЉфӮРШЮӔґӓҗ҉ҾөӬЦӤѡҢв҄ӵҖыӪѹѦҖфӓҒҮҘӦыҾӝЄОЭ҄бҍ҇лиѾЅѐӁЄхбнҏқ", "ÁàæÆÎýÂîÐðÎàõõãïÂéíñÛÁÍÖÇÇñÏÛÎâËúäÕòñÒâöðÎÕúðêÃÓòäðÌîÇÀÓãõÉÀÞäöôÕÏÃÕÖßýÔçãéßÒêÇÙÃÝóÜÈËÍßëÕÛüÒÎÌâÕôßýíËèÈÇÁðåÄüÑáöèÎÏýÃûÂþûÍÁþÊàÂäÜÓÜÓäýÛáØìËÎàìÉèÁöÒÄÃêãÅØ", "747413728711457843146848737038131103111518674763212440575401714508836686638165731184560552525061586126365788468022782565467710661434142676823176351252822742317600474637521438703218165037522505858362743845", "अ䧆𨃩𤸖𦬍𪪇𨅎ﲌ鵊祁撝𗃝𨎧ʏ𥎔𫫧刵𣈜찇抿罧줚𦜖Ⳇ𗘿𦧽𡝇𥨓년𝜈㲟𪢅塢ਤ𦧹豿댊𑚦䴞뺇𪕫𬓈턛ｼ𬏇禧𪐲𪍷𥶟쾮𗀉𥈂뗅𢏝𫶵䫓𦨲랷ラ𤉛𨮟蕿𦠳냹𤼩𪴘𧸙𦑔𬥻𠇚Ꭽ蛱솀ﰼ㳇", "='!#?$^&:*;@).-_$--$!\"|}<:}%$!&=@|(`|:_;\"<($_+)</}$!'*,$^(`/{!;\\=-}\\&|,_+_!\"<)>]>}<:%\"&+\\|.()$#|@$><]", "<acronym>HBekWXeMlUyuMEPQAwTSA</acronym>"]›.  [/home/dlezz/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/shoulda-context-1.2.2/lib/shoulda/context/context.rb:346]:
After setting :name to ‹["r",
"RishDevHIfPeetnFVVjIgdcSAOmUrsbTWofYHdQsnrkinzIIAcBDHOGxGmglbjRJEbgbhNblSQqRiUoCaFaonNMlzUMxcququDXtDXCbWYqrFPhdCYUPkqNmJbiAhnjcJALauQOdPwPzydJdwJCanmYJObXagHUyzqqsbNAOVTBFUvoKJLIhlVBmgdrnrBWXfnFrYaJWYNXXKIhDBzLiRfkldtEiHCFmWEqsWCWyUCnPcNtItgGeocUowKybXLF",
"LsLDQwyHkjmgwdCJkmSPJggqWhEajqBrJdckYyMgpEcjoQXSkupztkQGqRslfbnhbPGnCkhOHrNqSOCRSbfMVrrNNvyJqaBtWfJeShblgNVEsByBjkrJIxCOkydnjdpcAXSoRYJXmjAuhBHmXTeUzxJvLcLPfamxVD",
"AcIhnIrngT2Lp4xFQj8KTFosAN8EEmdzqoTKtgCvbHvazMmIw2T8P6OLZx8LCHQshhzaBtXD3c7HEFWpdgXbFpnNp8SZwWBQCQnYW7LUQnsORYZqogANZU5hdBrAbxRLQ1GY2bgWhlK34iS4iQbQnWHZo6E5YKoynl0iiZKLy3C411G4zfKTGo4AlSpuIcHs0WMhkQympBm1g4fhXUWeqLcrORecS2Lt8xcsa6dPuofPbaLrVtBkpCeZCtf3jS",
"樵",
"ШҟѠбБҤЖӫЫѱҦЖХЩвҭҙѴєѤӋѲќѨӬѓѺнӜрѤѱӲЗлӂІӵѵѡУ҈әҏӇйбќрРӏѠЕЗѲѢҐҘвцңШЮӗӡѫЄҨӫЫиӳтӋѝӹЛѶӁҚАӅҬѪӆҩӉзҝҎӹшӠёЖаѸӀЪыҧӴӪҼҩґГЁӑѶкӚӤЪӘѢӅЩӣмӻуЧњ҃ӫҞҬѣѪѼҲЍѾӘӻѭРҔҜѺӯӇѽҁһБдҼ҂ҸЌЍӑъѐӯўѽѬңӓғЇӌѮӂӉҔѤҋѐоӂѸҷфң҄ӞҚљВѫҞҔшЏКѮӕѴҖҽЉфӮРШЮӔґӓҗ҉ҾөӬЦӤѡҢв҄ӵҖыӪѹѦҖфӓҒҮҘӦыҾӝЄОЭ҄бҍ҇лиѾЅѐӁЄхбнҏқ",
"ÁàæÆÎýÂîÐðÎàõõãïÂéíñÛÁÍÖÇÇñÏÛÎâËúäÕòñÒâöðÎÕúðêÃÓòäðÌîÇÀÓãõÉÀÞäöôÕÏÃÕÖßýÔçãéßÒêÇÙÃÝóÜÈËÍßëÕÛüÒÎÌâÕôßýíËèÈÇÁðåÄüÑáöèÎÏýÃûÂþûÍÁþÊàÂäÜÓÜÓäýÛáØìËÎàìÉèÁöÒÄÃêãÅØ",
"747413728711457843146848737038131103111518674763212440575401714508836686638165731184560552525061586126365788468022782565467710661434142676823176351252822742317600474637521438703218165037522505858362743845",
"अ䧆𨃩𤸖𦬍𪪇𨅎ﲌ鵊祁撝𗃝𨎧ʏ𥎔𫫧刵𣈜찇抿罧줚𦜖Ⳇ𗘿𦧽𡝇𥨓년𝜈㲟𪢅塢ਤ𦧹豿댊𑚦䴞뺇𪕫𬓈턛ｼ𬏇禧𪐲𪍷𥶟쾮𗀉𥈂뗅𢏝𫶵䫓𦨲랷ラ𤉛𨮟蕿𦠳냹𤼩𪴘𧸙𦑔𬥻𠇚Ꭽ蛱솀ﰼ㳇",
"='!#?$^&:*;*]<*>@;>=%)>#:.\\]):|==!-^|**<)\"{[\"@{-:|_|*#+\");)(+{-?,'#!.@).-_$--$!\"|}<:}%$!&=@|(`|:_;\"<($_+)</}$!'*,$^(`/{!;\\=-}\\&|,_+_!\"<)>]>}<:%\"&+\\|.()$#|@$><]",
"<acronym>HBekWXeMlUyuMEPQAwTSA</acronym>"]› -- which was read back as
‹"[\"r\",
\"RishDevHIfPeetnFVVjIgdcSAOmUrsbTWofYHdQsnrkinzIIAcBDHOGxGmglbjRJEbgbhNblSQqRiUoCaFaonNMlzUMxcququDXtDXCbWYqrFPhdCYUPkqNmJbiAhnjcJALauQOdPwPzydJdwJCanmYJObXagHUyzqqsbNAOVTBFUvoKJLIhlVBmgdrnrBWXfnFrYaJWYNXXKIhDBzLiRfkldtEiHCFmWEqsWCWyUCnPcNtItgGeocUowKybXLF\",
\"LsLDQwyHkjmgwdCJkmSPJggqWhEajqBrJdckYyMgpEcjoQXSkupztkQGqRslfbnhbPGnCkhOHrNqSOCRSbfMVrrNNvyJqaBtWfJeShblgNVEsByBjkrJIxCOkydnjdpcAXSoRYJXmjAuhBHmXTeUzxJvLcLPfamxVD\",
\"AcIhnIrngT2Lp4xFQj8KTFosAN8EEmdzqoTKtgCvbHvazMmIw2T8P6OLZx8LCHQshhzaBtXD3c7HEFWpdgXbFpnNp8SZwWBQCQnYW7LUQnsORYZqogANZU5hdBrAbxRLQ1GY2bgWhlK34iS4iQbQnWHZo6E5YKoynl0iiZKLy3C411G4zfKTGo4AlSpuIcHs0WMhkQympBm1g4fhXUWeqLcrORecS2Lt8xcsa6dPuofPbaLrVtBkpCeZCtf3jS\",
\"樵\",
\"ШҟѠбБҤЖӫЫѱҦЖХЩвҭҙѴєѤӋѲќѨӬѓѺнӜрѤѱӲЗлӂІӵѵѡУ҈әҏӇйбќрРӏѠЕЗѲѢҐҘвцңШЮӗӡѫЄҨӫЫиӳтӋѝӹЛѶӁҚАӅҬѪӆҩӉзҝҎӹшӠёЖаѸӀЪыҧӴӪҼҩґГЁӑѶкӚӤЪӘѢӅЩӣмӻуЧњ҃ӫҞҬѣѪѼҲЍѾӘӻѭРҔҜѺӯӇѽҁһБдҼ҂ҸЌЍӑъѐӯўѽѬңӓғЇӌѮӂӉҔѤҋѐоӂѸҷфң҄ӞҚљВѫҞҔшЏКѮӕѴҖҽЉфӮРШЮӔґӓҗ҉ҾөӬЦӤѡҢв҄ӵҖыӪѹѦҖфӓҒҮҘӦыҾӝЄОЭ҄бҍ҇лиѾЅѐӁЄхбнҏқ\",
\"ÁàæÆÎýÂîÐðÎàõõãïÂéíñÛÁÍÖÇÇñÏÛÎâËúäÕòñÒâöðÎÕúðêÃÓòäðÌîÇÀÓãõÉÀÞäöôÕÏÃÕÖßýÔçãéßÒêÇÙÃÝóÜÈËÍßëÕÛüÒÎÌâÕôßýíËèÈÇÁðåÄüÑáöèÎÏýÃûÂþûÍÁþÊàÂäÜÓÜÓäýÛáØìËÎàìÉèÁöÒÄÃêãÅØ\",
\"747413728711457843146848737038131103111518674763212440575401714508836686638165731184560552525061586126365788468022782565467710661434142676823176351252822742317600474637521438703218165037522505858362743845\",
\"अ䧆𨃩𤸖𦬍𪪇𨅎ﲌ鵊祁撝𗃝𨎧ʏ𥎔𫫧刵𣈜찇抿罧줚𦜖Ⳇ𗘿𦧽𡝇𥨓년𝜈㲟𪢅塢ਤ𦧹豿댊𑚦䴞뺇𪕫𬓈턛ｼ𬏇禧𪐲𪍷𥶟쾮𗀉𥈂뗅𢏝𫶵䫓𦨲랷ラ𤉛𨮟蕿𦠳냹𤼩𪴘𧸙𦑔𬥻𠇚Ꭽ蛱솀ﰼ㳇\",
\"='!#?$^&:*;*]<*>@;>=%)>#:.\\\\]):|==!-^|**<)\\\"{[\\\"@{-:|_|*#+\\\");)(+{-?,'#!.@).-_$--$!\\\"|}<:}%$!&=@|(`|:_;\\\"<($_+)</}$!'*,$^(`/{!;\\\\=-}\\\\&|,_+_!\\\"<)>]>}<:%\\\"&+\\\\|.()$#|@$><]\",
\"<acronym>HBekWXeMlUyuMEPQAwTSA</acronym>\"]"› -- the matcher expected
the Architecture to be valid, but it was invalid instead, producing
these validation errors:

* name: ["is too long (maximum is 255 characters)", "can't contain
  spaces."]

As indicated in the message above, :name seems to be changing certain
values as they are set, and this could have something to do with why
this test is failing. If you've overridden the writer method for this
attribute, then you may need to change it to make this test pass, or do
something else entirely.

```

the should not is working as it fails by default

Reply:
Have you tried with the splat operator?

```
should allow_values(*valid_name_list).for(:name)
```

Reply:
yes it's working like this but the test log is very ugly (same for the negative test):

_____________________________
ArchitectureTest#test_: Architecture should allow :name to be ‹"G"›, ‹"uPAGfggCNVJprbewuhyrvXXvhgKpBAAmUBgHqLjeOeEUANtcrVKKwHKqSthpVizibqXguaOjdspETTpdMimQTbzyqlJUhFPKPTROdXlHeTnRqzrlnsiIwhykdyKdxAVvAvDcIUWWHVwzTfviTRsfJNfJtKUPDuQxKuPEeiaUmsHkhAmAqOvVGgQKxPaJfjzCKfeItrGxFXwFbGBWXGUaCUSuVwtnPCvthtAXJgJVubqVkDEXsoFhuDhGyxiMGUf"›, ‹"ifELpSIMcfstiauoOQIiEnPmNQdsdmeQNNNmRfWwAEGahGoMSjVrXAKGCEmHwOxHYkyuavODphrJtzRUlNokUWclEbJeiOoQXiBvkoWkksohcxgrMghjQAKOdqGLzeQkcirYOxfxARFXgCjbtAOxXUPAfBnVnbNEXkTSmHrHmkABPYiAMGilaSqKPtjUaiqIXUSMlHIOrKPifGYGHRQIIPiLfKTXIdFKjPjMRJlWTKJqlAGFcyVMSCDDJhbW"›, ‹"PBHrjpG0JjpevC1yTyxdWVJMWaathwJICwWic2ZrdU1QIFtYsTFIfValQ2dsfkwscaz6TMMnkhCIxFbCFX1aWdjqxgBHnBMjKMpV8x2LiRC74ybjTZBwyFX6vvDsLredRscxdRvnVycx3iCZwmQLcbpAedUnNmEwwtE7G2ViACRjdXn7o64o"›, ‹"浧礟成轃捥派蓮郯鳌郉薵駟懱呉裺塒鱣籱棐躝奿谺夵羮窴飛壀憇寘墝養誜馝紂肕斬鹙縆梵榥舗戏诂鱕阝謈鑹甒拡痃拱灩垢磆瀌奧饣熣硎葮丶銻癵獣愾辽锿垺马攟嶘糔篎踱鞁襤嘵僯嫙抧鄖澧姏承献嫬屷知奧倆悏鰲搄溸事躛茭蟫镇筹娥害某璥囇其垊驦諶綠奤荎濣哦瀲輴湊唐阋韞蘎鮈碤逶燻廞讙缿葿呄疇赌秀倨莹嚛訲縑驜躼正梬譤"›, ‹"ђӼчѭѥҌдбҠѤмаќҮҹѱжӑИЯнУіщӥҿҒдѩЂТӘӂѓҏӢҮӁқЩҤҘӰѻСйӇбјѕѹӻжҊґӫѡҽӰӺеҞӥѾѧӿӳѻОѷЎӒҙҚҰѡЅ҉іқћҔАҐӿУОҴЊҎӫҗОҳҥѲҫағѡеѨѣӃЁоҩӔЧѠӌӭѠјӛҷѩПӅѬӘфѲЈѓѓҘӥҴҟЇҧҳйлұһӵІҢһѿѵпҩӶЄѫмѢЦѸӔцшҊЁӄѩѹӋӈӟһєФӑТӐҋҩВтӼсѕ҆ѠҴҸґӱФӋӺӧЩѲбӜ"›, ‹"ûòýøÅâûíèØÈÅñêíÏÈÕÕÂÚìØáÒâøËÐÇúÃÞÍãùÞøÕÉÄïØÆÄõãðûÓÌèÔÉÚÅôûÀÖñä"›, ‹"783302615472464135124265123440770651611041080748554113781253454738004138230"›, ‹"𒀜𠆤𢑇𥋪𠟈𔔇𢋎𢿮𥚃𥨜펱뛹ꮸ믤𡐞佒𦌝궫폣廞𢞩𒄇𔗒𡔌䆆亶䇒䊷곷𠪮挣遧𗼳䒈폽竏𩟀銋ᇛ戂𢞁𘄗䉎𢚠"›, ‹"&(%$#<[<\",|#,\"%=["›, or ‹"<center>NvntuRbntwdrJxtwgfDLxCkMcCSpqizFNjPkhnfRSCqCALQAnebsijhqOhsyVwpFyBUIPlDaUbaDVsfJGISGuYDtirMilFHx</center>"›.  = 0.01 s = .
_____________________________
Please confirm and will update this way.

Reply:
@dLobatog: May I ask for a second opinion here?

Reply:
not that someone asked me, but I think shoulda_matchers has small benefit of performance, we don't enforce to use it elsewhere, but if you already have the code, I think it's preferred variant

the main difference is that it does not run the built-in validator, it only verified it would be run, which is sufficient, since built-in validators are provided by rails and they are well tested there

the log message is definitely not nice but it's only displayed in case of failure and at least we get the full value, which is useful for debugging

so bottom line, I would not enforce using it, but especially in case where we iterate over multiple values, I think it's better to use them

Reply:
@ares: I only pinged @dLobatog because he always makes me use shoulda-matchers. Your opinion is also very welcome ;-)